### PR TITLE
Assignment7+8 - LINQ - Ethan Guerin and Rachael Deboer

### PR DESCRIPTION
**Pull request by EthanAGi**
Rachael's Git - @rkdeboer 
Ethan's Git - @EthanAGi

### CODE REVIEW

**Comment by colestewart12** [comment]
I can review!

File: Assignment.Tests/SampleData.Tests.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        // Sample Data added to the test file
+        var testCsvContent = "FirstName,LastName,StreetAddress,City,State,Zip,EmailAddress\n" +
+                             "John,Doe,123 Main St,Anytown,CA,12345,john@example.com\n" +
+                             "Jane,Smith,456 Elm St,Other City,NY,67890,jane@example.com";
+        File.WriteAllText(TestCsvFilePath, testCsvContent);
+    }
+
+    [TestCleanup]
+    public void Cleanup()
```
Comment (colestewart12) [inline Assignment.Tests/SampleData.Tests.cs:26]: Nice job with cleanup!

**Review by colestewart12 — COMMENTED** [review]
## Instructions
Throughout, consider using the System.Linq.Enumerable methods Zip, Count, Sort and Contains methods for testing collections.. (Preferably avoid using Microsoft.VisualStudio.TestTools.UnitTesting.CollectionAssert although that might be easier, to get a firmer grasp on additional LINQ API.)

Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ❌
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ❌✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.
## Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌
## Fundamentals
Place all shared project properties into a Directory.Build.Props file.✔
Place all shared project items into a Directory.Build.targets file.❌
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Overall looks pretty good! Might be able to simplify a couple of the LINQ commands, nice work!

File: Assignment/Assignment/Assignment.csproj [hunk 2]
```diff
@@ -4,4 +4,10 @@
     <ProjectReference Include="..\Assignment.Interfaces\Assignment.Interfaces.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Update="People.csv">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
```
[the remark this answers is not in the document]
Comment (rkdeboer) [reply Assignment/Assignment/Assignment.csproj:9]: Thank you!

File: Assignment/Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,102 @@
...
     {
-        // 1.
-        public IEnumerable<string> CsvRows => throw new NotImplementedException();
+        get 
+        {
+            if (!File.Exists(CsvFilePath))
+            {
+                throw new FileNotFoundException(nameof(CsvFilePath));
+            }
+            return File.ReadLines(CsvFilePath).Skip(1);// This should return the entire list, Line by Line, as a string
```
[the remark this answers is not in the document]
Comment (rkdeboer) [reply Assignment/Assignment/SampleData.cs:22]: Thank you! 😄

File: Assignment/Assignment/SampleData.cs [hunk 4]
```diff
@@ -1,30 +1,102 @@
...
+                                     //splitting it at the 5th comma, and ordering by state
 
-        // 3.
-        public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+    // 3.
+    public string GetAggregateSortedListOfStatesUsingCsvRows()
+    {
+        var sortedUniqueStates = GetUniqueSortedListOfStatesGivenCsvRows()
+            .ToArray();
```
[the remark this answers is not in the document]
Comment (rkdeboer) [reply Assignment/Assignment/SampleData.cs:41]: Thank you!

File: Assignment/Assignment/SampleData.cs [hunk 5]
```diff
@@ -1,30 +1,102 @@
...
 
-        // 4.
-        public IEnumerable<IPerson> People => throw new NotImplementedException();
+        return string.Join(",", sortedUniqueStates);
+    }
+
+    // 4.
```
[the remark this answers is not in the document]
Comment (rkdeboer) [reply Assignment/Assignment/SampleData.cs:46]: Thank you!

File: Assignment/Assignment/SampleData.cs [hunk 6]
```diff
@@ -1,30 +1,102 @@
...
+            })
+    .Where(person => person != null) //This gets rid of null entries
+    .OrderBy(p => p.Address.State)
+    .ThenBy(p => p.Address.City)
+    .ThenBy(p => p.Address.Zip)
+    .Cast<IPerson>();
+        }
+    }
+
+    // 5.
```
[the remark this answers is not in the document]
Comment (rkdeboer) [reply Assignment/Assignment/SampleData.cs:77]: Thank you!

File: Assignment/Assignment/SampleData.cs [hunk 7]
```diff
@@ -1,30 +1,102 @@
...
+        Predicate<string> filter)
+    {
+        return filter == null
+            ? throw new ArgumentNullException(nameof(filter))
+            : People.Where(p => p.EmailAddress != null && filter(p.EmailAddress))
+            .Select(p => (p.FirstName, p.LastName));
+    } 
+
+    // 6.
+    public string GetAggregateListOfStatesGivenPeopleCollection(
```
[the remark this answers is not in the document]
Comment (rkdeboer) [reply Assignment/Assignment/SampleData.cs:88]: Thank you!

File: Assignment.Tests/SampleData.Tests.cs (hunk 1 above)
Comment (rkdeboer) [reply Assignment.Tests/SampleData.Tests.cs:26]: Thank you!

File: Assignment.Tests/Assignment.Tests.csproj [hunk 8]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <IsPackable>false</IsPackable>
+	  <TargetFramework>.net8.0</TargetFramework>
```
Comment (montalcy) [inline Assignment.Tests/Assignment.Tests.csproj:5]: Dont forget to set your lang version

File: Assignment.Tests/Node.Tests.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+{
+    [TestMethod]
+    public void Append_SingleItem_Succeed()
+    {
+        Node<int> node = new(1);
+        node.Append(2);
+        var expected = new[] { 1, 2 };
+        var actual = node.ToArray();
+
+        CollectionAssert.AreEqual(expected, actual);
```
Comment (montalcy) [inline Assignment.Tests/Node.Tests.cs:17]: I wouldnt have thought to use CollectionAssert, nice!

File: Assignment.Tests/Person.Tests.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Assignment;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+namespace Assignment.Tests;
+public class PersonTests
+{
+    [TestMethod]
```
Comment (montalcy) [inline Assignment.Tests/Person.Tests.cs:10]: I didnt think to test the classes that we're given to us, but good double check

File: Assignment.Tests/SampleData.Tests.cs [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+using System.IO;
+using System.Linq;
+
+namespace Assignment.Tests;
+
+[TestClass]
+public class SampleDataTests
+{
+    //I tried my best with these but only 1 is currently passing haha
+    private const string TestCsvFilePath = "People.csv"; // Makes a testing file with sample data
```
Comment (montalcy) [inline Assignment.Tests/SampleData.Tests.cs:13]: We were having a hard time creating sample data as well. Nice job!

File: Assignment.Tests/SampleData.Tests.cs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+        Assert.AreEqual(2, result.Count);
+        CollectionAssert.Contains(result, "John,Doe,123 Main St,Anytown,CA,12345,john@example.com");
+        CollectionAssert.Contains(result, "Jane,Smith,456 Elm St,Other City,NY,67890,jane@example.com");
+    }
+
+    [TestMethod]
+    public void GetUniqueSortedListOfStatesGivenCsvRows_ReturnsUniqueSortedEmails()
+    {
+        SampleData sampleData = new();
```
Comment (montalcy) [inline Assignment.Tests/SampleData.Tests.cs:48]: Do you think you could just reuse sampleData between tests to save some time?

File: Assignment.Tests/Person.Tests.cs (hunk 10 above)
Comment (rkdeboer) [reply Assignment.Tests/Person.Tests.cs:10]: You're probably right, I just have all this testing stuff drilled into me now, leave no line untouched haha!

File: Assignment.Tests/SampleData.Tests.cs (hunk 11 above)
Comment (rkdeboer) [reply Assignment.Tests/SampleData.Tests.cs:13]: It was a task but we got it done!

**Review by montalcy — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ❌
Implement ISampleDate.FilterByEmailAddress(Predicate filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5
Implement IEnumerable to return all the items in the "circle" of items. ✔
Add an IEnumberable ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.✔

Fundamentals
Place all shared project properties into a Directory.Build.Props file.✔
Place all shared project items into a Directory.Build.targets file.❌ -- I didnt see a targets file
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)  
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Good job yall. I left a few questions in the comments as well, but overall well done.

File: Assignment/Assignment/SampleData.cs [hunk 13]
```diff
@@ -1,30 +1,102 @@
...
+    // 2.
+    public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+        => CsvRows
+            .Select(row => row.Split(','))
+            .Where(fields => fields.Length > 4)
+            .Select(fields => fields[4].Trim())
+            .Where(state => !string.IsNullOrEmpty(state))
+            .Distinct()
+            .OrderBy(state => state);//This will use CsvRows, calling in one large string,
+                                     //splitting it at the 5th comma, and ordering by state
```
[the remark this answers is not in the document]
Comment (rkdeboer) [reply Assignment/Assignment/SampleData.cs]: I see what you are saying but I think this may be missing a step. I tried what you had here and it seems to have broken things. The aggregated lists tests began failing. I am not sure what the missing part is but I appreciate the suggestion!

File: Assignment.Tests/Assignment.Tests.csproj (hunk 8 above)
Comment (EthanAGi) [reply Assignment.Tests/Assignment.Tests.csproj:5]: Thank you! Just fixed this!

File: Assignment.Tests/SampleData.Tests.cs (hunk 12 above)
Comment (rkdeboer) [reply Assignment.Tests/SampleData.Tests.cs]: Good point! I just added that change! Thanks!

File: Assignment/Assignment/SampleData.cs [hunk 14]
```diff
@@ -1,30 +1,102 @@
...
+        return string.Join(",", sortedUniqueStates);
+    }
+
+    // 4.
+    public IEnumerable<IPerson> People
+    {
+        get
+        {
+            return CsvRows.Select(row => row.Split(','))
+                .Where(fields => fields.Length >= 7)
```
[the remark this answers is not in the document]
Comment (EthanAGi) [reply Assignment/Assignment/SampleData.cs:52]: Removed!

File: Assignment.Tests/Assignment.Tests.csproj [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/Assignment.Tests.csproj:5]:
```suggestion
	  <TargetFramework>net8.0</TargetFramework>
```
bug

File: Assignment.Tests/Assignment.Tests.csproj [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+	  <LangVersion>latest</LangVersion>
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/Assignment.Tests.csproj:6]: This is set in the directory.build.props file

File: Assignment.Tests/Node.Tests.cs (hunk 9 above)
Comment (BenjaminMichaelis) [reply Assignment.Tests/Node.Tests.cs:17]: While not a dock of points, the direction did say 

> "Preferably avoid using `CollectionAssert` although that might be easier, to get a firmer grasp on additional LINQ API."

File: Assignment.Tests/Node.Tests.cs [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    [TestMethod]
+    public void Exists_ItemExists_ReturnsTrue()
+    {
+        Node<int> node = new(1);
+        node.Append(2);
+        node.Append(3);
+
+        Assert.IsTrue(node.Exists(2));
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/Node.Tests.cs:39]: `Assert.Contains` should handle this case if it's built properly like a collection
```
        Assert.Contains(node, 2);
```

File: Assignment.Tests/Node.Tests.cs [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        Node<int> node = new(1);
+        node.Append(2);
+        node.Append(3);
+
+        Assert.IsFalse(node.Exists(4));
+    }
+
+    [TestMethod]
+    public void ChildItems_ReturnsCorrect_Succeed()
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/Node.Tests.cs:53]: What about edge cases? What if I give it a max of `-4`? Should it throw an exception, should it return empty? Good things to think about. There could also be a good argument made for whether maximum should be inclusive or exclusive and could be good to test for.

File: Assignment.Tests/SampleData.Tests.cs [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+        Assert.AreEqual("CA", result[0]);
+        Assert.AreEqual("NY", result[1]);
+    }
+
+    [TestMethod]
+    public void GetAggregateSortedListOfStatesUsingCsvRows_ReturnsAggregateSortedStates()
+    {
+
+
+        var result = sampleData.GetAggregateSortedListOfStatesUsingCsvRows();
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/SampleData.Tests.cs:62]: nit, throughout this file: whitespace
```suggestion
    {
        var result = sampleData.GetAggregateSortedListOfStatesUsingCsvRows();
```

File: Assignment/Assignment/SampleData.cs [hunk 20]
```diff
@@ -1,30 +1,100 @@
...
+                state: fields[4],
+                zip: fields[5]),
+            emailAddress: fields[6]
+        );
 
-        // 3.
-        public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+            })
+    .Where(person => person != null) //This gets rid of null entries
```
Comment (BenjaminMichaelis) [inline Assignment/Assignment/SampleData.cs:67]: Annoying thing with where is that nullability isn't changed... :)

It's why i've made extension methods like WhereNotNull - https://github.com/IntelliTect/Multitool?tab=readme-ov-file#extensions

File: Assignment/Directory.Build.props
Comment (BenjaminMichaelis) [inline Assignment/Directory.Build.props:1]: Why isn't this at the root of your solution?

File: Assignment/Assignment/SampleData.cs [hunk 21]
```diff
@@ -1,30 +1,100 @@
...
+            }
+            return File.ReadLines(CsvFilePath).Skip(1);// This should return the entire list, Line by Line, as a string
+        }
+    }
+
+    // 2.
+    public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+       => CsvRows
+            .Select(row => row.Split(','))
+            .Where(fields => fields.Length > 4)
```
Comment (BenjaminMichaelis) [inline Assignment/Assignment/SampleData.cs:30]: Some comments could make this code more clear to read
